templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-sources.tsx

```diff
@@ -23,82 +23,150 @@ function SourceNumberButton({ index }: { index: number }) {
 type NodeInfo = {
   id: string;
   url?: string;
+  text?: string;
+  file_name?: string;
+  metadata?: Record<string, unknown>;
 };
 
-export function ChatSources({ data }: { data: SourceData }) {
-  const sources: NodeInfo[] = useMemo(() => {
-    // aggregate nodes by url or file_path (get the highest one by score)
-    const nodesByPath: { [path: string]: NodeInfo } = {};
+type Document = {
+  url: string;
+  sources: NodeInfo[];
+};
+
+function isValidUrl(url: string): boolean {
+  try {
+    new URL(url);
+    return true;
+  } catch (_) {
+    return false;
+  }
+}
 
+export function ChatSources({ data }: { data: SourceData }) {
+  const documents: Document[] = useMemo(() => {
+    // group nodes by url
+    const nodesByUrl: { [path: string]: NodeInfo[] } = {};
```

Comment:
The code below is trying to use Url and than Id. Can't we always group by Id?

Reply:
I mean not using the node ID, but the ID of the document. It's usually called doc_id, or ref_id. Can you have a look and pass it?